File: finddup.py
#!/usr/bin/python3
import os
import sys

filelist = {}
lastfilesize = None
lastfilename = None

def sha1sum(file):
	import hashlib

	BLOCKSIZE = 65536
	hasher = hashlib.sha1()
	with open(file, 'rb') as afile:
		buf = afile.read(BLOCKSIZE)
		while len(buf) > 0:
			hasher.update(buf)
			buf = afile.read(BLOCKSIZE)

	return hasher.hexdigest()

def processdir(path, filelist):
	for filename in os.listdir(path):
		fullpath = os.path.join(path, filename)
		if os.path.isdir(fullpath):
			# further down the rabbit hole you go
			processdir(fullpath, filelist)
		elif os.path.isfile(fullpath):
			filelist[fullpath] = os.path.getsize(fullpath)

for directory in sys.argv[1:]:
	if os.path.isdir(directory):
		processdir(directory, filelist)

for filename, filesize in sorted(filelist.items(), key=lambda x:x[1]):
	# being the same size is a necessary, but not sufficient condition
	if filesize == lastfilesize:
		if sha1sum(filename) == sha1sum(lastfilename):
			print (str(filename) + " and " + str(lastfilename) + " are duplicates")

	lastfilename = filename
	lastfilesize = filesize

File: finddup.php
<?php
$filelist = array();
$lastfilesize = -1;
$lastfilename = '';

function joinpath() {
    $paths = array();

    foreach (func_get_args() as $arg)
        if ($arg !== '')
		$paths[] = $arg;

    return preg_replace('#/+#','/',join('/', $paths));
}

function processdir($dir, &$filelist)
{
	if ($directoryhandler = opendir($dir))
		while ($file = readdir($directoryhandler))
			if (($file != '.') and ($file != '..'))
			{
				$fullpath = joinpath($dir, $file);
				if (is_file($fullpath) AND (!is_link($fullpath)))
					$filelist[$fullpath] = filesize($fullpath);
				elseif (is_dir($fullpath))
					// further down the rabbit hole you go
					processdir($fullpath, $filelist);
			}
	closedir($directoryhandler);
}

foreach (array_slice($argv,1) as $directory)
{
	if (is_dir($directory))
		processdir($directory, $filelist);
}

asort($filelist, SORT_NUMERIC);
foreach ($filelist as $filename => $filesize)
{
	// being the same size is a necessary, but not sufficient condition
	if ($filesize === $lastfilesize)
		if (sha1_file($filename) === sha1_file($lastfilename))
			echo $filename . " and " . $lastfilename . " are duplicates\n";

	$lastfilename = $filename;
	$lastfilesize = $filesize;
}
?>
